<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#damage_zones(-)-to-calculate-parameters-of-damage-zones" data-toc-modified-id="damage_zones(-)-to-calculate-parameters-of-damage-zones-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>damage_zones( ) to calculate parameters of damage zones</a></span></li><li><span><a href="#impact_risk(-)-uncertainty-analysis" data-toc-modified-id="impact_risk(-)-uncertainty-analysis-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>impact_risk( ) uncertainty analysis</a></span></li></ul></div>

To install the module and any pre-requisites, from the base directory run:
```
pip install -r requirements.txt
pip install -e .
```

To download the postcode data required:

```
python download_data.py
```

In [19]:
from armageddon.damage import damage_zones, impact_risk
from armageddon.solver import Planet
import armageddon
import folium

### damage_zones( ) to calculate parameters of damage zones

Calculate the latitude and longitude of the surface zero location and the
list of airblast damage radii (m) for a given impact scenario.


__Parameters__

outcome: Dict

    the outcome dictionary from an impact scenario
        
lat: float

    latitude of the meteoroid entry point (degrees)

lon: float

    longitude of the meteoroid entry point (degrees)
    
bearing: float

    Bearing (azimuth) relative to north of meteoroid trajectory (degrees)
    
pressures: float, arraylike

    List of threshold pressures to define airblast damage levels
    
plot: bool

    Boolean value to decide plotting

__Return__

blat: float

    latitude of the surface zero point (degrees)
        
blon: float

    longitude of the surface zero point (degrees)
        
damrad: arraylike, float

    List of distances specifying the blast radii for the input damage levels
        
plot: plot object

    The plot specifying the areas effected by each damage level


__Examples__
    
    >>> import armageddon
    >>> outcome = {'burst_altitude': 8e3, 'burst_energy': 7e3,
                   'burst_distance': 90e3, 'burst_peak_dedz': 1e3,
                   'outcome': 'Airburst'}
    >>> armageddon.damage_zones(outcome, 52.79, -2.95, 135,
                                pressures=[1e3, 3.5e3, 27e3, 43e3])\


In [41]:
outcome = {'burst_altitude': 8e3, 'burst_energy': 7e3,
           'burst_distance': 90e3, 'burst_peak_dedz': 1e3,
           'outcome': 'Airburst'}
arr = damage_zones(outcome, 52.79, -2.95, 135, 
                   pressures=[1e3, 3.5e3, 27e3, 43e3],
                   map = True)
arr

(51.98371949678459,
 -2.746093100475223,
 [51002.04019315775,
  25706.014907978413,
  7196.208816950395,
  3803.4573273729134])

_Explaination:_

damage_zones( ) returned a list of parameters with :

    surface zero latitude in degree:  51.98371949678459

    surface zero longitude in degree:  -2.746093100475223

    blast radii with the pressure level of 1e3:  51002.04019315775

    blast radii with the pressure level of 3.5e3:  25706.014907978413

    blast radii with the pressure level of 27e3:  7196.208816950395

    blast radii with the pressure level of 43e3:  3803.4573273729134


In [35]:
a = armageddon.plot_circle(arr[0], arr[1], arr[2][0])
for i in range(1, len(arr[2])):
    armageddon.plot_circle(arr[0], arr[1], arr[2][i], map = a)

In [34]:
a = armageddon.plot_circle(arr[0], arr[1], arr[2][0])
armageddon.plot_circle(arr[0], arr[1], arr[2][1], map = a)
armageddon.plot_circle(arr[0], arr[1], arr[2][2], map = a)
armageddon.plot_circle(arr[0], arr[1], arr[2][3], map = a)

_Explaination:_
    
    The damage zones are marked on the plot above. As the plot and calculations shows, as the radius(horizontal range) goes up, the pressure in this wave decreases. This is also proved by the equation:
    
\begin{equation*}
p(r) = 3.14 \times 10^{11} \left(\frac{r^2 + z_b^2}{E_k^{2/3}}\right)^{-1.3} + 1.8 \times 10^{7} \left(\frac{r^2 + z_b^2}{E_k^{2/3}}\right)^{-0.565}
\end{equation*}

### impact_risk( ) uncertainty analysis

Perform an uncertainty analysis to calculate the risk for each affected UK postcode or postcode sector


__Parameters__

planet: armageddon.Planet instance

    The Planet instance from which to solve the atmospheric entry

means: dict

    A dictionary of mean input values for the uncertainty analysis. This should include values for ``radius``, ``angle``, ``strength``, ``density``, ``velocity``, ``lat``, ``lon`` and ``bearing``

stdevs: dict

    A dictionary of standard deviations for each input value. This should include values for ``radius``, ``angle``, ``strength``, ``density``, ``velocity``, ``lat``, ``lon`` and ``bearing``

pressure: float

    A single pressure at which to calculate the damage zone for each impact

nsamples: int

    The number of iterations to perform in the uncertainty analysis

sector: logical, optional

    If True (default) calculate the risk for postcode sectors, otherwise calculate the risk for postcodes

__Return__

risk: DataFrame

    A pandas DataFrame with columns for postcode (or postcode sector) and the associated risk. These should be called ``postcode`` or ``sector``, and ``risk``.


In [9]:
p = Planet()
fiducial_means = {'radius': 35, 'angle': 45, 'strength': 1e7,
                  'density': 3000, 'velocity': 19e3,
                  'lat': 53.0, 'lon': -2.5, 'bearing': 115.}
fiducial_stdevs = {'radius': 1, 'angle': 1, 'strength': 5e6,
                   'density': 500, 'velocity': 1e3,
                   'lat': 0.025, 'lon': 0.025, 'bearing': 0.5}
impact_risk(p, means=fiducial_means, stdevs=fiducial_stdevs, pressure=27.e3, nsamples=10, sector=True)

[['DE6 2' 'DE6 1' 'ST10 3' 'B79 0' 'LE65 1' 'CV13 0' 'LE67 2' 'DE12 7'
  'LE65 2' 'DE6 3' 'CV13 6' 'ST10 4' 'ST14 5' 'DE12 6' 'LE67 6' 'DE6 5'
  'SK17 0' 'CV9 3' 'DE12 8' 'DE6 9' 'SK13 5' 'SK14 6' 'SK2 5' 'SK4 5'
  'SK8 2' 'SK8 5' 'SK7 0' 'SK5 8' 'SK4 2' 'SK13 6' 'SK11 0' 'SK22 1'
  'SK6 1' 'SK22 4' 'SK10 3' 'SK7 1' 'SK10 2' 'SK11 8' 'SK14 3' 'SK6 4'
  'SK3 8' 'SK9 2' 'SK6 0' 'ST14 8' 'SK11 6' 'SK7 6' 'SK4 3' 'SK3 0'
  'SK9 4' 'SK9 6' 'SK6 5' 'SK23 9' 'SK4 1' 'SK8 1' 'SK10 5' 'SK8 6'
  'SK11 9' 'SK3 9' 'B77 2' 'SK2 7' 'B77 3' 'SK1 3' 'ST14 9' 'SK12 1'
  'SK3 3' 'SK6 2' 'SK6 9' 'SK2 9' 'SK8 7' 'SK10 9' 'SK5 7' 'SK1 2'
  'DE65 5' 'SK6 3' 'SK14 5' 'SK5 9' 'B79 9' 'SK22 2' 'SK1 4' 'B78 1'
  'SK6 7' 'SK7 4' 'SK9 0' 'LE67 1' 'SK14 1' 'B78 2' 'B77 5' 'CV9 1'
  'SK9 3' 'SK23 0' 'SK9 7' 'CV9 9' 'B77 4' 'SK9 1' 'SK13 8' 'SK1 9'
  'SK7 5' 'CV9 2' 'SK10 1' 'SK2 6' 'SK22 3' 'ST13 7' 'ST14 7' 'SK12 2'
  'M34 7' 'SK23 7' 'SK1 1' 'SK22 9' 'SK6 8' 'SK7 3' 'SK9 5' 'SK23 6'
  'SK6 6' 'SK17 6' 'SK17 9' 'S

IndexError: index 0 is out of bounds for axis 0 with size 0